In [1]:
import numpy as np
import pandas as pd 

In [1]:
df = pd.read_csv('train.csv',usecols=['X1','X2','X3','X4','X5','X6'])

In [6]:
df.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [10]:
#how many labels each variable has
for col in df.columns:
    print(col,':' ,len(df[col].unique()),'labels')

X1 : 27 labels
X2 : 44 labels
X3 : 7 labels
X4 : 4 labels
X5 : 29 labels
X6 : 12 labels


In [11]:
for col in df.columns:
    print(col,':' ,df[col].value_counts(),'labels')

X1 : aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
w      52
z      46
u      37
e      33
m      32
t      31
h      29
f      23
y      23
j      22
n      19
k      17
p       9
g       6
d       3
q       3
ab      3
Name: X1, dtype: int64 labels
X2 : as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
z       19
ag      19
d       18
ac      13
g       12
y       11
ap      11
x       10
aw       8
at       6
h        6
al       5
q        5
an       5
p        4
av       4
ah       4
au       3
o        1
l        1
af       1
c        1
ar       1
j        1
am       1
aa       1
Name: X2, dtype: int64 labels
X3 : c    1942
f    1076
a     440
d     290
g     241
e     163
b      57
Name: X3, dtype: int64 labels
X4 : d    4205
a       2
b       1
c       1
Name: X4, dtype: int64 lab

In [12]:
#lets examine the how many columns we will obtain after one hot encoding
pd.get_dummies(df,drop_first=True).shape

(4209, 117)

we can see that from 6 initial categorical variables,we end up with 117 new variables

In [13]:
#lets find the top 10 frequent categories for the variable x2

df.X2.value_counts().sort_values(ascending=False).head(20)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
z       19
ag      19
Name: X2, dtype: int64

In [17]:
#lets make a list with most frequent categories of the variable
top_10 = [x for x in df.X2.value_counts().sort_values(ascending=False).head(10).index]

In [18]:
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [20]:
#now we make the 10 binary variables

for label in top_10:
    df[label] = np.where(df['X2']==label,1,0)
df[['X2']+top_10].head()

X2  as  ae  ai  m  ak  r  n  s  f  e
0  at   0   0   0  0   0  0  0  0  0  0
1  av   0   0   0  0   0  0  0  0  0  0
2   n   0   0   0  0   0  0  1  0  0  0
3   n   0   0   0  0   0  0  1  0  0  0
4   n   0   0   0  0   0  0  1  0  0  0

In [25]:
#get whole set of dummy variables,for all the categorical variables

def one_hot_top_x(data,variable,top_x_labels):
    for label in top_x_labels:
        data[variable+'_'+label]=np.where(df[variable]==label,1,0)
        
df = pd.read_csv('train.csv',sep=",", usecols=['X1','X2','X3','X4','X5','X6'])

#encode x2 into 10 most frequent categories
one_hot_top_x(df,'X2',top_10)
df.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0  v  at  a  d  u  j      0      0      0     0      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0      0     0     1     0   
3  t   n  f  d  x  l      0      0      0     0      0     0     1     0   
4  v   n  f  d  h  d      0      0      0     0      0     0     1     0   

   X2_f  X2_e  
0     0     0  
1     0     0  
2     0     0  
3     0     0  
4     0     0

In [26]:
top_10 = [x for x in df.X1.value_counts().sort_values(ascending=False).head(10).index]
one_hot_top_x(df,'X1',top_10)
df.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X1_aa  X1_s  X1_b  X1_l  \
0  v  at  a  d  u  j      0      0      0     0  ...      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...      0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...      0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...      0     0     0     0   

   X1_v  X1_r  X1_i  X1_a  X1_c  X1_o  
0     1     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     0     0     0     0  
3     0     0     0     0     0     0  
4     1     0     0     0     0     0  

[5 rows x 26 columns]